ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/truncate/TruncateTableAnalyzer.java

```diff
@@ -338,4 +370,56 @@ private void addStatTask(ASTNode root, Table table, Path oldPartitionLocation, P
       moveTask.addDependentTask(statTask);
     }
   }
+
+  public StringBuilder constructDeleteQuery(Table table, Map<String, String> partitionSpec) throws SemanticException {
+    String qualifiedTableName = HiveTableName.ofNullable(HiveUtils.unparseIdentifier(table.getTableName(), this.conf),
+            HiveUtils.unparseIdentifier(table.getDbName(), this.conf), null).getNotEmptyDbTable();
+    StringBuilder sb = new StringBuilder().append("delete from ").append(qualifiedTableName)
+            .append(" where ");
+    Deserializer deserializer = table.getDeserializer();
+    Map<String, PrimitiveObjectInspector.PrimitiveCategory> stringTypeInfoMap = new HashMap<>();
+    try {
+      ObjectInspector objectInspector = deserializer.getObjectInspector();
+      if (objectInspector.getCategory() == ObjectInspector.Category.STRUCT) {
+        StructObjectInspector structObjectInspector = (StructObjectInspector) objectInspector;
+        List<? extends StructField> structFields =  structObjectInspector.getAllStructFieldRefs();
+        for (StructField structField : structFields) {
+          if (structField.getFieldObjectInspector().getCategory() == ObjectInspector.Category.PRIMITIVE) {
+            PrimitiveObjectInspector primitiveObjectInspector = (PrimitiveObjectInspector) structField.getFieldObjectInspector();
+            stringTypeInfoMap.put(structField.getFieldName(),
+                primitiveObjectInspector.getTypeInfo().getPrimitiveCategory());
+          }
+        }
+      }
+    } catch (SerDeException e) {
+      throw new SemanticException(String.format("Unable to get object inspector due to: %s", e));
+    }
+    boolean first = true;
+    for (String key : partitionSpec.keySet()) {
+      PrimitiveObjectInspector.PrimitiveCategory category = stringTypeInfoMap.get(key);
+      if (category == null) {
+        throw new SemanticException(String.format(
+            "Only primitive partition column type is supported via TRUNCATE operation " +
+            " but column %s is not a primitive category.", key));
+      }
+      String value = partitionSpec.get(key);
+      boolean shouldEncloseQuotes = TypeInfoUtils.shouldEncloseQuotes(category);
+      if (value != null) {
+        sb.append(first ? "" : " and ").append(key).append(" = ");
```

Comment:
i'll leave this to @kasakrisz as i think there is a better way to quote all column names